napari/_vispy/experimental/tiled_image_visual.py

```diff
@@ -0,0 +1,354 @@
+"""TiledImageVisual class
+
+A visual that draws tiles using a texture atlas.
+"""
+from typing import List, Set
+
+import numpy as np
+from vispy.gloo.buffer import VertexBuffer
+
+from ...layers.image.experimental.octree_util import ChunkData
+from ..vendored import ImageVisual
+from ..vendored.image import _build_color_transform
+from .texture_atlas import TextureAtlas2D
+from .tile_set import TileSet
+
+# Shape of she whole texture in tiles. Hardcode for now.
+SHAPE_IN_TILES = (16, 16)
+
+
+# Two triangles to cover a [0..1, 0..1] quad.
+_QUAD = np.array(
+    [[0, 0], [1, 0], [1, 1], [0, 0], [1, 1], [0, 1]], dtype=np.float32,
+)
+
+
+def _vert_quad(chunk_data: ChunkData) -> np.ndarray:
+    """Return quad for the vertex buffer.
+
+    Parameters
+    ----------
+    chunk_data : ChunkData
+        Create a quad for this chunk.
+
+    Return
+    ------
+    np.darray
+        The quad vertices.
+    """
+    quad = _QUAD.copy()
+
+    # TODO_OCTREE: store as np.array in ChunkData?
+    scale = np.array(chunk_data.scale, dtype=np.float32)
+    scaled_shape = chunk_data.data.shape[:2] * scale
+
+    # Modify XY's into place
+    quad[:, :2] *= scaled_shape
+    quad[:, :2] += chunk_data.pos
+
+    return quad
+
+
+def _tex_quad(chunk_data: ChunkData) -> np.ndarray:
+    """Return quad for the texture coordinate buffer.
+
+    Parameters
+    ----------
+    chunk_data : ChunkData
+        Create a quad for this chunk.
+
+    Return
+    ------
+    np.darray
+        The quad texture coordinates.
+    """
+    quad = _QUAD.copy()[:, :2]
+
+    # TODO_OCTREE: store as np.array in ChunkData?
+    scale = np.array(chunk_data.scale, dtype=np.float32)
+    scaled_shape = chunk_data.data.shape[:2] * scale
+
+    # Modify XY's into place
+    quad[:, :2] *= scaled_shape
+    quad[:, :2] += chunk_data.pos
+
+    return quad
+
+
+class TiledImageVisual(ImageVisual):
```

Comment:
Yeah the weird thing is our model code cannot depend on vispy. Our VispyVisuals can depend on model code. But then vispy itself cannot depend on any of our code at all.

The level_index should not be in the `ChunkData` that vispy knows about. So then it's just data/pos/scale. So you can just pass those in as 3 variables. And have vispy have a private namedtuple or dataclass for them.

The level_index is more complicated. Vispy should not have the concept of a level, that's octree specific. But vispy would need some type of generic identifier so callers can signal if a chunk is really the same one or not. Not really sure how that would look. It could be a hash of the data and the pos/scale, but hashing data is slower than some other things.

Reply:
Yeah, sounds like we're not too far away, but some details would need to be worked out. We can revisit once we have everything working.